neo/Consensus/ConsensusService.cs

```diff
@@ -112,7 +112,7 @@ private void CheckExpectedView(byte view_number)
 
         private void CheckPreparations()
         {
-            if (context.Preparations.Count(p => p) >= context.M && context.TransactionHashes.All(p => context.Transactions.ContainsKey(p)))
+            if (context.Preparations.Count(p => p != null) >= context.M && context.TransactionHashes.All(p => context.Transactions.ContainsKey(p)))
```

Comment:
Shouldn’t this be checking that `context.M` Count of the Perparations equal the primary’s hash instead of just `!= null`.

Reply:
I think that this is ensured in a previous layer, when receiving the Response/Request or when saving the variable if Response arrived before Request.

Reply:
Fully agree with you @jsolman ... tested or not, comparing to null is very similar to comparing to hash, which will look much safer.

Reply:
But that is what I meant in my comment above, I still prefer our first design with signature instead of Hash. We were all the time necessarily checking them.

Reply:
@igormcoelho Yes, i think it works in most of the cases even if this wasn't changed to my suggestion, but there is one case I'm not so sure about when it needs to get another transaction. I think we agree though it is a nice safeguard to change it as I suggest here, and could only do more help than harm.

Reply:
@vncoelho I actually think even without making this change the code is actually safe though, since it always checks for `ConsensusState.RequestReceived` and when that is received it null's out all the Preparations that don't match the primary, so this isn't really an issue if merged the way it is, but it looks safer and is almost as efficient to just make it as I suggest.

Reply:
https://github.com/neo-project/neo/blob/9ce07a5c22e0d275fdc5c7004320595533874e73/neo/Consensus/ConsensusService.cs#L253-L256

When we are ready to send a `Commit` message, we must have received the `PrepareRequest`. So at this time, the values of hashes in `context.Preparations` must be equal.

Reply:
@erikzhang

I agree, but it doesn’t hurt to explicitly check.